Src/FluentAssertions/Common/TypeExtensions.cs

```diff
@@ -136,6 +136,34 @@ public static bool IsEquivalentTo(this IMember property, IMember otherProperty)
 
         public static Type[] GetClosedGenericInterfaces(Type type, Type openGenericType)
         {
+            // Avoid expensive calculation when type cannot possibly implement the interface we
+            // care about. The only TypeCodes that can implement interfaces are Object and String,
```

Comment:
This is not universally true.
E.g. `int` has `TypeCode.Int32` and implements the generic interfaces `IComparable<int>` and `IEquatable<int>`.

Reply:
Huh, yeah, that's true... I think there are cross purposes here -- on the one hand, we're trying to make the best possible `GetClosedGenericInterfaces` method, while on the other hand we're only ever using it to find implementations of 
`IEnumerable<T>` and `IDictionary<TKey, TValue>`.

But, in keeping with the first goal, perhaps the thing to do is to refactor this, creating a separate `GetClosedGenericInterfacesFast` method that checks all of the non-object `TypeCode` values.

Reply:
While residing in `TypeExtensions` `GetClosedGenericInterfaces` should not make any assumptions about who's calling it.

If the performance penalty is from `CanHandle` we could even have something like a `HasClosedGenericInterface(Type type, Type openGenericType)` to avoiding enumerating all interfaces.

```
|                               Method |           Job |       Runtime |             Type |      Mean |     Error |   StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------------- |-------------- |-------------- |----------------- |----------:|----------:|---------:|------:|--------:|-------:|------:|------:|----------:|
|           GetClosedGenericInterfaces |    .NET 4.7.2 |    .NET 4.7.2 |   IEnumerable<T> |  41.40 ns |  0.547 ns | 0.485 ns |  1.00 |    0.00 | 0.0140 |     - |     - |      88 B |
|           HasClosedGenericInterfaces |    .NET 4.7.2 |    .NET 4.7.2 |   IEnumerable<T> |  19.40 ns |  0.209 ns | 0.196 ns |  0.47 |    0.01 | 0.0038 |     - |     - |      24 B |
| HasClosedGenericInterfacesImperative |    .NET 4.7.2 |    .NET 4.7.2 |   IEnumerable<T> |  15.41 ns |  0.175 ns | 0.164 ns |  0.37 |    0.01 |      - |     - |     - |         - |
|                ImplementsIEnumerable |    .NET 4.7.2 |    .NET 4.7.2 |   IEnumerable<T> |  91.13 ns |  1.357 ns | 1.203 ns |  2.20 |    0.03 |      - |     - |     - |         - |
|                                      |               |               |                  |           |           |          |       |         |        |       |       |           |
|           GetClosedGenericInterfaces | .NET Core 3.1 | .NET Core 3.1 |   IEnumerable<T> |  37.53 ns |  0.458 ns | 0.406 ns |  1.00 |    0.00 | 0.0140 |     - |     - |      88 B |
|           HasClosedGenericInterfaces | .NET Core 3.1 | .NET Core 3.1 |   IEnumerable<T> |  17.76 ns |  0.373 ns | 0.458 ns |  0.47 |    0.02 | 0.0038 |     - |     - |      24 B |
| HasClosedGenericInterfacesImperative | .NET Core 3.1 | .NET Core 3.1 |   IEnumerable<T> |  13.31 ns |  0.276 ns | 0.258 ns |  0.35 |    0.00 |      - |     - |     - |         - |
|                ImplementsIEnumerable | .NET Core 3.1 | .NET Core 3.1 |   IEnumerable<T> |  57.57 ns |  0.744 ns | 0.659 ns |  1.53 |    0.02 |      - |     - |     - |         - |
|                                      |               |               |                  |           |           |          |       |         |        |       |       |           |
|           GetClosedGenericInterfaces |    .NET 4.7.2 |    .NET 4.7.2 | IEnumerable<int> |  81.03 ns |  1.575 ns | 1.875 ns |  1.00 |    0.00 | 0.0139 |     - |     - |      88 B |
|           HasClosedGenericInterfaces |    .NET 4.7.2 |    .NET 4.7.2 | IEnumerable<int> |  55.21 ns |  0.781 ns | 0.730 ns |  0.68 |    0.02 | 0.0038 |     - |     - |      24 B |
| HasClosedGenericInterfacesImperative |    .NET 4.7.2 |    .NET 4.7.2 | IEnumerable<int> |  51.07 ns |  0.792 ns | 0.741 ns |  0.63 |    0.01 |      - |     - |     - |         - |
|                ImplementsIEnumerable |    .NET 4.7.2 |    .NET 4.7.2 | IEnumerable<int> |  95.48 ns |  1.932 ns | 2.147 ns |  1.18 |    0.05 |      - |     - |     - |         - |
|                                      |               |               |                  |           |           |          |       |         |        |       |       |           |
|           GetClosedGenericInterfaces | .NET Core 3.1 | .NET Core 3.1 | IEnumerable<int> |  60.20 ns |  1.192 ns | 1.276 ns |  1.00 |    0.00 | 0.0140 |     - |     - |      88 B |
|           HasClosedGenericInterfaces | .NET Core 3.1 | .NET Core 3.1 | IEnumerable<int> |  36.59 ns |  0.712 ns | 0.631 ns |  0.60 |    0.02 | 0.0038 |     - |     - |      24 B |
| HasClosedGenericInterfacesImperative | .NET Core 3.1 | .NET Core 3.1 | IEnumerable<int> |  31.61 ns |  0.640 ns | 0.657 ns |  0.52 |    0.02 |      - |     - |     - |         - |
|                ImplementsIEnumerable | .NET Core 3.1 | .NET Core 3.1 | IEnumerable<int> |  62.71 ns |  1.244 ns | 1.222 ns |  1.04 |    0.03 |      - |     - |     - |         - |
|                                      |               |               |                  |           |           |          |       |         |        |       |       |           |
|           GetClosedGenericInterfaces |    .NET 4.7.2 |    .NET 4.7.2 |          List<T> | 515.38 ns |  7.338 ns | 6.864 ns |  1.00 |    0.00 | 0.0553 |     - |     - |     353 B |
|           HasClosedGenericInterfaces |    .NET 4.7.2 |    .NET 4.7.2 |          List<T> | 271.24 ns |  4.548 ns | 4.254 ns |  0.53 |    0.01 | 0.0329 |     - |     - |     209 B |
| HasClosedGenericInterfacesImperative |    .NET 4.7.2 |    .NET 4.7.2 |          List<T> | 211.41 ns |  4.145 ns | 4.435 ns |  0.41 |    0.01 | 0.0138 |     - |     - |      88 B |
|                ImplementsIEnumerable |    .NET 4.7.2 |    .NET 4.7.2 |          List<T> | 211.78 ns |  4.194 ns | 3.923 ns |  0.41 |    0.01 |      - |     - |     - |         - |
|                                      |               |               |                  |           |           |          |       |         |        |       |       |           |
|           GetClosedGenericInterfaces | .NET Core 3.1 | .NET Core 3.1 |          List<T> | 385.82 ns |  6.674 ns | 6.243 ns |  1.00 |    0.00 | 0.0548 |     - |     - |     344 B |
|           HasClosedGenericInterfaces | .NET Core 3.1 | .NET Core 3.1 |          List<T> | 196.23 ns |  3.892 ns | 4.780 ns |  0.50 |    0.01 | 0.0331 |     - |     - |     208 B |
| HasClosedGenericInterfacesImperative | .NET Core 3.1 | .NET Core 3.1 |          List<T> | 137.95 ns |  2.664 ns | 3.272 ns |  0.36 |    0.01 | 0.0138 |     - |     - |      88 B |
|                ImplementsIEnumerable | .NET Core 3.1 | .NET Core 3.1 |          List<T> | 144.93 ns |  2.851 ns | 4.354 ns |  0.37 |    0.01 |      - |     - |     - |         - |
|                                      |               |               |                  |           |           |          |       |         |        |       |       |           |
|           GetClosedGenericInterfaces |    .NET 4.7.2 |    .NET 4.7.2 |        List<int> | 578.55 ns | 10.642 ns | 9.954 ns |  1.00 |    0.00 | 0.0553 |     - |     - |     353 B |
|           HasClosedGenericInterfaces |    .NET 4.7.2 |    .NET 4.7.2 |        List<int> | 320.92 ns |  6.308 ns | 7.747 ns |  0.56 |    0.02 | 0.0329 |     - |     - |     209 B |
| HasClosedGenericInterfacesImperative |    .NET 4.7.2 |    .NET 4.7.2 |        List<int> | 254.49 ns |  5.003 ns | 6.327 ns |  0.44 |    0.01 | 0.0138 |     - |     - |      88 B |
|                ImplementsIEnumerable |    .NET 4.7.2 |    .NET 4.7.2 |        List<int> | 211.15 ns |  3.239 ns | 2.871 ns |  0.37 |    0.01 |      - |     - |     - |         - |
|                                      |               |               |                  |           |           |          |       |         |        |       |       |           |
|           GetClosedGenericInterfaces | .NET Core 3.1 | .NET Core 3.1 |        List<int> | 411.51 ns |  7.994 ns | 8.554 ns |  1.00 |    0.00 | 0.0548 |     - |     - |     344 B |
|           HasClosedGenericInterfaces | .NET Core 3.1 | .NET Core 3.1 |        List<int> | 197.59 ns |  2.484 ns | 2.323 ns |  0.48 |    0.01 | 0.0331 |     - |     - |     208 B |
| HasClosedGenericInterfacesImperative | .NET Core 3.1 | .NET Core 3.1 |        List<int> | 162.54 ns |  3.261 ns | 2.891 ns |  0.40 |    0.01 | 0.0138 |     - |     - |      88 B |
|                ImplementsIEnumerable | .NET Core 3.1 | .NET Core 3.1 |        List<int> | 133.03 ns |  2.058 ns | 1.925 ns |  0.32 |    0.01 |      - |     - |     - |         - |
|                                      |               |               |                  |           |           |          |       |         |        |       |       |           |
|           GetClosedGenericInterfaces |    .NET 4.7.2 |    .NET 4.7.2 |              int | 281.33 ns |  5.340 ns | 5.244 ns |  1.00 |    0.00 | 0.0367 |     - |     - |     233 B |
|           HasClosedGenericInterfaces |    .NET 4.7.2 |    .NET 4.7.2 |              int | 211.38 ns |  4.072 ns | 4.847 ns |  0.75 |    0.03 | 0.0293 |     - |     - |     185 B |
| HasClosedGenericInterfacesImperative |    .NET 4.7.2 |    .NET 4.7.2 |              int | 149.19 ns |  2.998 ns | 2.945 ns |  0.53 |    0.01 | 0.0100 |     - |     - |      64 B |
|                ImplementsIEnumerable |    .NET 4.7.2 |    .NET 4.7.2 |              int |  92.91 ns |  1.813 ns | 1.607 ns |  0.33 |    0.01 |      - |     - |     - |         - |
|                                      |               |               |                  |           |           |          |       |         |        |       |       |           |
|           GetClosedGenericInterfaces | .NET Core 3.1 | .NET Core 3.1 |              int | 196.77 ns |  3.945 ns | 4.385 ns |  1.00 |    0.00 | 0.0331 |     - |     - |     208 B |
|           HasClosedGenericInterfaces | .NET Core 3.1 | .NET Core 3.1 |              int | 155.42 ns |  2.375 ns | 2.222 ns |  0.79 |    0.02 | 0.0305 |     - |     - |     192 B |
| HasClosedGenericInterfacesImperative | .NET Core 3.1 | .NET Core 3.1 |              int |  99.70 ns |  1.971 ns | 2.827 ns |  0.51 |    0.02 | 0.0114 |     - |     - |      72 B |
|                ImplementsIEnumerable | .NET Core 3.1 | .NET Core 3.1 |              int |  58.60 ns |  0.223 ns | 0.186 ns |  0.30 |    0.01 |      - |     - |     - |         - |
|                                      |               |               |                  |           |           |          |       |         |        |       |       |           |
|           GetClosedGenericInterfaces |    .NET 4.7.2 |    .NET 4.7.2 |           string | 380.22 ns |  6.556 ns | 5.812 ns |  1.00 |    0.00 | 0.0548 |     - |     - |     345 B |
|           HasClosedGenericInterfaces |    .NET 4.7.2 |    .NET 4.7.2 |           string | 228.03 ns |  3.078 ns | 2.728 ns |  0.60 |    0.01 | 0.0317 |     - |     - |     201 B |
| HasClosedGenericInterfacesImperative |    .NET 4.7.2 |    .NET 4.7.2 |           string | 159.31 ns |  2.762 ns | 2.449 ns |  0.42 |    0.01 | 0.0126 |     - |     - |      80 B |
|                ImplementsIEnumerable |    .NET 4.7.2 |    .NET 4.7.2 |           string | 214.73 ns |  2.785 ns | 2.605 ns |  0.57 |    0.01 |      - |     - |     - |         - |
|                                      |               |               |                  |           |           |          |       |         |        |       |       |           |
|           GetClosedGenericInterfaces | .NET Core 3.1 | .NET Core 3.1 |           string | 289.26 ns |  5.239 ns | 4.644 ns |  1.00 |    0.00 | 0.0534 |     - |     - |     336 B |
|           HasClosedGenericInterfaces | .NET Core 3.1 | .NET Core 3.1 |           string | 113.16 ns |  1.581 ns | 1.401 ns |  0.39 |    0.01 | 0.0318 |     - |     - |     200 B |
| HasClosedGenericInterfacesImperative | .NET Core 3.1 | .NET Core 3.1 |           string |  64.99 ns |  1.168 ns | 1.035 ns |  0.22 |    0.00 | 0.0126 |     - |     - |      80 B |
|                ImplementsIEnumerable | .NET Core 3.1 | .NET Core 3.1 |           string | 137.32 ns |  1.453 ns | 1.359 ns |  0.48 |    0.01 |      - |     - |     - |         - |
```

```cs
[Benchmark(Baseline = true)]
public bool GetClosedGenericInterfaces() => GetClosedGenericInterfaces(Type.Type, typeof(IEnumerable<>)).Any();

[Benchmark]
public bool HasClosedGenericInterfaces() => HasClosedGenericInterfaces(Type.Type, typeof(IEnumerable<>));

[Benchmark]
public bool HasClosedGenericInterfacesImperative() => HasClosedGenericInterfacesImperative(Type.Type, typeof(IEnumerable<>));

[Benchmark]
public bool ImplementsIEnumerable() => ImplementsIEnumerable(Type.Type);

public static bool HasClosedGenericInterfaces(Type type, Type openGenericType)
{
    return type.IsGenericType && type.GetGenericTypeDefinition() == openGenericType
        || type.GetInterfaces().Any(t => t.IsGenericType && t.GetGenericTypeDefinition() == openGenericType);
}

public static bool HasClosedGenericInterfacesImperative(Type type, Type openGenericType)
{
    if (type.IsGenericType && type.GetGenericTypeDefinition() == openGenericType)
    {
        return true;
    }

    foreach (var t in type.GetInterfaces())
    {
        if (t.IsGenericType && t.GetGenericTypeDefinition() == openGenericType)
        {
            return true;
        }
    }

    return false;
}

public static Type[] GetClosedGenericInterfaces(Type type, Type openGenericType)
{
    return type.IsGenericType && type.GetGenericTypeDefinition() == openGenericType
        ? new[] { type }
        : type.GetInterfaces()
            .Where(t => t.IsGenericType && t.GetGenericTypeDefinition() == openGenericType)
            .ToArray();
}

public static bool ImplementsIEnumerable(Type type)
{
    var @interface = type.GetInterface("IEnumerable`1");
    return @interface == typeof(IEnumerable<>)
        || @interface?.IsGenericType == true && @interface.GetGenericTypeDefinition() == typeof(IEnumerable<>);
}
```

Reply:
I have added a commit to this end: 865f9639802b4f6f0082c1b03d322f6d1f7dc695

Reply:
The optimization in this area was not based on in-depth profiling, but on the observation that across a couple of dozen debug sessions, almost every time I paused execution during a run that was taking abnormally run, it was inside `type.GetInterfaces()` on a primitive type, inside a call stack that was checking if it implemented `IEnumerable<>` or `IDictionary<,>`. After the optimization, I observed the runtime improve significantly. I did not profile it rigorously before or after, though, so I don't have specific numbers to back it up :-)